feat(ssrf): adds rules to check for IP based SSRF

fzipi opened the pull request:
- extends the original IPs handled in RFI rules to IPv6 and other IPv4 evasions
- added the regular expressions used to data file
- tests check the expressions

Signed-off-by: Felipe Zipitria <felipe.zipitria@owasp.org>

dune73 commented:
I'd like to see this into the rules file that #2163 renames.

fzipi commented:
Is this is a generic attack now, or is it because two rules are not too much to throw in new files?

fzipi commented:
What about the new `ssrf_score`? Does it make sense to people here?

dune73 commented:
I'm personally not a fan of the separate scores, but some people are. In this particular case I doubt it makes much sense, unless it's really a separate rule group.

Generic or not: There are a lot of different rule categories. Adding a new rule category / group brings at least 8 administrative rules. That's a lot if it's only going to be 2-3 rules. So I think we need a group for "various rules". In my eyes, the new generic rules group was such a place.

franbuehler commented:
[Meeting decision December](https://github.com/coreruleset/coreruleset/issues/2291#issuecomment-987122691): This PR will be shifted into the new category "generic attacks" after #2163 is merged.

fzipi commented:
SSRF rules, 🎄 edition! 🎉 

- moved rules to "generic attacks"
- added all tests for each bypass

Before merge, we just need to think:
- right now it is applying only to `ARGS`: do we need additional variables?
- any other transformation to prevent evasions?

franbuehler commented:
[Meeting decision January](https://github.com/coreruleset/coreruleset/issues/2330#issuecomment-1004310622): @franbuehler will review it and pass it to @lifeforms for testing.

fzipi commented:
We're you able to review this one @franbuehler ?

franbuehler commented:
Hi @fzipi,
Sorry for my late reply. I didn't see your ping.
No, unfortunately not yet...

franbuehler commented on tests/regression/tests/REQUEST-934-APPLICATION-ATTACK-GENERIC/934120.yaml:
```diff
@@ -0,0 +1,692 @@
+            output:
+              no_log_contains: id "934120"
+    -
+      test_title: 934120-21
```
We have two -21 tests.

fzipi360 replied:
Good catch! Fixed.

franbuehler commented on util/regexp-assemble/data/934120.data:
```diff
@@ -0,0 +1,67 @@
+##! This is a data file used to generate a regular expression for a CRS rule.
```
I was not able to generate the regexp. But maybe it's just me??!

I get this error. What I'm doing wrong?

If it's just me, it's ok.

```
$ python3 regexp-assemble.py data/934120.data 
Traceback (most recent call last):
  File "regexp-assemble.py", line 149, in <module>
    run()
  File "regexp-assemble.py", line 146, in run
    assemble(lines)
  File "regexp-assemble.py", line 133, in assemble
    sys.stdout.write(outs.decode('utf-8'))
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe2 in position 1312: invalid continuation byte
```

fzipi replied:
Hmmm...
```
❯ python3 regexp-assemble.py data/934120.data
(?i)(?:acap|bitcoin|blob|cap|cvs|svn|svn\+ssh|turn|udp|vnc|xmpp|webcal|xri|dav|dns|feed|finger|h323|icap|ipp|ipps|jabber|mms|mumble|maven|nfs|rmi|rsync|rtmp|rtm|rtmfp|http|https|ftp|ftps|git|gopher|irc|fd|ldap|ldapi|ldaps|expect|netdoc|jar|ogg|tftp|dict|ssh2|ssh|imap|pop2|pop3|mailto|nntp|news|snews|snmp|smtp|sftp|telnet|s3|phar|sip|sips|smb)://(?:\[(?:((?:[0-9a-f]{1,4}:){7,7}[a-f0-9]{1,4}|(?:[a-f0-9]{1,4}:){1,7}:|(?:[a-f0-9]{1,4}:){1,6}:[a-f0-9]{1,4}|(?:[a-f0-9]{1,4}:){1,5}(?::[a-f0-9]{1,4}){1,2}|(?:[a-f0-9]{1,4}:){1,4}(?::[a-f0-9]{1,4}){1,3}|(?:[a-f0-9]{1,4}:){1,3}(?::[a-f0-9]{1,4}){1,4}|(?:[a-f0-9]{1,4}:){1,2}(?::[a-f0-9]{1,4}){1,5}|[a-f0-9]{1,4}:(?:(?::[a-f0-9]{1,4}){1,6})|:(?:(?::[a-f0-9]{1,4}){1,7}|:)|fe80:(?::[a-f0-9]{0,4}){0,4}%[a-f0-9]{1,}|::(?:ffff(:0{1,4}){0,1}:){0,1}(?:(?:25[0-5]|(?:2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(?:25[0-5]|(?:2[0-4]|1{0,1}[0-9]){0,1}[0-9])|(?:[a-f0-9]{1,4}:){1,4}:(?:(25[0-5]|(?:2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(?:25[0-5]|(?:2[0-4]|1{0,1}[0-9]){0,1}[0-9]))|(?:(?:[0-9a-f]{1,4}:){6}(?:(25[0-5]|(?:2[0-4]|1{0,1}\d){0,1}\d)\.){3,3}(?:25[0-5]|(?:2[0-4]|1{0,1}\d){0,1}\d))\])|\xe2(?:\x93(?:\x81\xe2\x93\x84\xe2\x92\xb8\xe2\x92\xb6\xe2\x93\x81\xe2\x92\xbd\xe2\x93\x84\xe2\x93\x88\xe2\x93\x89|\x9b\xe2\x93\x9e\xe2\x93\x92\xe2\x93\x90\xe2\x93\x9b\xe2\x93\x97\xe2\x93\x9e\xe2\x93\xa2\xe2\x93\xa3)|\x91\xa0\xe2\x91(?:\xa1\xe2\x91\xa6|\xa5\xe2\x91\xa8)\.)|[a-zA-Z][\w\-\.]{1,255}:\d{1,5}\+&@(?:\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}|[a-zA-Z][\w\-\.]{1,255}):\d{1,5}\/?#\+@(?:\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}|[a-zA-Z][\w\-\.]{1,255}):\d{1,5}\/?|(?:0x[a-f0-9]{2}\.){3}0x[a-f0-9]{2}|(?:0{1,4}\d{1,3}\.){3}0{1,4}\d{1,3}|\d{1,3}\.(?:\d{1,3}\.\d{5}|\d{8})|0x(?:[a-f0-9]{16}|[a-f0-9]{8})|\d{10})
```

franbuehler replied:
Then it's just me... So ready to be tested by @lifeforms. PR looks good.

franbuehler reviewed (approved):
If it's just me getting this error while generating the regex (I wanted to check the output against the rule), this MR is ready to be tested by @lifeforms. 
Nice rule, a nice extension, thank you!

lifeforms commented:
I will run the rule on traffic starting next Wednesday.

dune73 commented:
Can you give us an estimate on the number of requests for this?

lifeforms commented:
I should improve my logging, but if I look at some access_logs from yesterday, I think something like 2 million requests.

dune73 commented:
Thank you.

And I can recommend this format: https://www.netnea.com/cms/apache-tutorial-5_extending-access-log/#step_4_configuring_the_new,_extended_log_format

```LogFormat "%h %{GEOIP_COUNTRY_CODE}e %u [%{%Y-%m-%d %H:%M:%S}t.%{usec_frac}t] \"%r\" %>s %b \
\"%{Referer}i\" \"%{User-Agent}i\" \"%{Content-Type}i\" %{remote}p %v %A %p %R \
%{BALANCER_WORKER_ROUTE}e %X \"%{cookie}n\" %{UNIQUE_ID}e %{SSL_PROTOCOL}x %{SSL_CIPHER}x \
%I %O %{ratio}n%% %D %{ModSecTimeIn}e %{ApplicationTime}e %{ModSecTimeOut}e \
%{ModSecAnomalyScoreInPLs}e %{ModSecAnomalyScoreOutPLs}e \
%{ModSecAnomalyScoreIn}e %{ModSecAnomalyScoreOut}e" extended```

Comes with a handy set of aliases too: https://raw.githubusercontent.com/Apache-Labor/labor/master/bin/.apache-modsec.alias

lifeforms commented:
Merged 🍴 because this rule is cool 😎 and I've got no false positives. (Sadly also no true positives)
